s3fs/core.py

```diff
@@ -1311,6 +1311,49 @@ async def _copy_basic(self, path1, path2, **kwargs):
             raise ValueError("Copy failed (%r -> %r): %s" % (path1, path2, e)) from e
         self.invalidate_cache(path2)
 
+    async def _copy_etag_preserved(self, path1, path2, size, total_parts, **kwargs):
+        """Copy file between locations on S3 as multi-part while preserving
+        the etag (using the same part sizes for each part"""
+
+        bucket1, key1, version1 = self.split_path(path1)
+        bucket2, key2, version2 = self.split_path(path2)
+
+        mpu = await self._call_s3(
+            self.s3.create_multipart_upload, Bucket=bucket2, Key=key2, **kwargs
+        )
+
+        parts = []
+        brange_first = 0
+        for i in range(1, total_parts + 1):
+            part_info = await self._call_s3(
+                self.s3.head_object, Bucket=bucket1, Key=key1, PartNumber=i
+            )
+            part_size = part_info["ContentLength"]
+            brange_last = brange_first + part_size - 1
+            if brange_last > size:
+                brange_last = size - 1
+
+            part = await self._call_s3(
```

Comment:
Can these be done concurrently with `asyncio.gather`?

Reply:
It seems like we can initially create futures for `head_object` calls and then `asyncio.gather` them and then another round for creating the upload calls and then gathering. So it seems doable

Reply:
Up to you on this one, please let me know if you want to stick with the original version here or if you think you can make `gather` work.

Reply:
I'll give it a shot later today, and let you know!

Reply:
It seems like using asyncio.gather on upload isn't a good option
https://github.com/dask/s3fs/blob/1333a8d46d38fd8d4c85536972cafcba998fd62e/s3fs/core.py#L1327-L1328
so I just did it for head object calls

Reply:
OK, that's fine. My guess would be that this happens because AWS doesn't know which chunk will be last. In any case, each call is quick, since the transfer payload is small.